src/perception/aruco_processing.cpp

```diff
@@ -2,85 +2,88 @@
 
 #include "filter.hpp"
 
-// TODO: add cache logic to filter out false positives
-
 /**
- * Detect fiducials from raw image using OpenCV and calculate their screen space centers.
- * Maintain the immediate buffer - holds all fiducials seen currently on screen.
- * Later camera space pose information is filled in when we receive point cloud data.
+ * Detect tags from raw image using OpenCV and calculate their screen space centers.
+ * Tag pose information relative to the camera in 3D space is filled in when we receive point cloud data.
  *
  * @param msg
  */
 void FiducialsNode::imageCallback(sensor_msgs::ImageConstPtr const& msg) {
     if (!mEnableDetections) return;
 
-    if (mIsVerbose) {
-        ROS_INFO("Got image %d", msg->header.seq);
-    }
+    ROS_DEBUG("Got image %d", msg->header.seq);
 
     try {
-        mCvPtr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
+        cv_bridge::CvImagePtr imagePtr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
 
-        // Detect the fiducial vertices in screen space and their respective ids
-        cv::aruco::detectMarkers(mCvPtr->image, mDictionary, mCorners, mIds, mDetectorParams);
+        // Detect the tag vertices in screen space and their respective ids
+        // {mCorners, mIds} are the outputs from OpenCV
+        cv::aruco::detectMarkers(imagePtr->image, mDictionary, mCorners, mIds, mDetectorParams);
 
-        // Remove fiducials in the immediate buffer that are no longer in sight
-        auto it = mImmediateFiducials.begin();
-        while (it != mImmediateFiducials.end()) {
-            if (std::find(mIds.begin(), mIds.end(), it->first) == mIds.end()) {
-                it = mImmediateFiducials.erase(it);
-            } else {
-                ++it;
-            }
-        }
+        ROS_DEBUG("OpenCV detect size: %zu", mIds.size());
 
-        // Save fiducials currently on screen to the immediate buffer
+        // Update ID, image center, and increment hit count for all detected tags
         for (size_t i = 0; i < mIds.size(); ++i) {
             int id = mIds[i];
-            cv::Point2f imageCenter = std::accumulate(mCorners[i].begin(), mCorners[i].end(), cv::Point2f{}) / 4.0f;
-            ImmediateFiducial& immediateFid = mImmediateFiducials[id];
-            immediateFid.id = id;
-            immediateFid.imageCenter = imageCenter;
+            Tag& tag = mTags[id];
+            tag.hitCount++;
```

Comment:
i'm thinking you'll want to cap the hit count. Having seen a tag 500 times before followed by not seeing it for many frames probably shouldn't give you any more confidence than having seen it 50 times before followed by not seeing it for many frames. Numbers are made up, you pick the params yourself.

Reply:
Yeah lol I just added that